tests/elasticity_test.py

```diff
@@ -0,0 +1,487 @@
+
+"""Tests for google3.third_party.py.jax_md.dynamics."""
+
+import numpy as onp
+
+from absl.testing import absltest
+from absl.testing import parameterized
+
+from jax.config import config as jax_config
+from jax import random
+from jax import jit
+from jax import lax
+from jax import grad
+import jax.numpy as np
+
+from jax import test_util as jtu
+
+from jax_md import space
+from jax_md import energy 
+from jax_md import minimize
+from jax_md import quantity
+from jax_md import elasticity
+from jax_md import test_util
+from jax_md.util import *
+
+jax_config.parse_flags_with_absl()
+jax_config.enable_omnistaging()
+FLAGS = jax_config.FLAGS
+
+PARTICLE_COUNT = 64
+STOCHASTIC_SAMPLES = 1
+SPATIAL_DIMENSION = [2, 3]
+
+LOWPRESSURE = [True, False]
+
+if FLAGS.jax_enable_x64:
+  #DTYPE = [f64]
+  DTYPE = [f32, f64]
+else:
+  DTYPE = [f32]
+
+def run_minimization_while(energy_fn, R_init, shift, max_grad_thresh = 1e-12, max_num_steps=1000000, **kwargs):
+  init,apply=minimize.fire_descent(jit(energy_fn), shift, **kwargs)
+  apply = jit(apply)
+
+  @jit
+  def get_maxgrad(state):
+    return jnp.amax(jnp.abs(state.force))
+
+  @jit
+  def cond_fn(val):
+    state, i = val
+    return jnp.logical_and(get_maxgrad(state) > max_grad_thresh, i<max_num_steps)
+
+  @jit
+  def body_fn(val):
+    state, i = val
+    return apply(state), i+1
+
+  state = init(R_init)
+  state, num_iterations = lax.while_loop(cond_fn, body_fn, (state, 0))
+
+  return state.position, get_maxgrad(state), num_iterations
+
+
+class ElasticityTest(jtu.JaxTestCase):
+  # pylint: disable=g-complex-comprehension
+  
+  @parameterized.named_parameters(jtu.cases_from_list(
+      {
+          'testcase_name': '_dim={}_dtype={}_lowpressure={}'.format(dim, dtype.__name__, low_pressure),
+          'spatial_dimension': dim,
+          'dtype': dtype,
+          'low_pressure': low_pressure
+      } for dim in SPATIAL_DIMENSION for dtype in DTYPE for low_pressure in LOWPRESSURE))
+  def test_EMT_from_database_basic(self, spatial_dimension, dtype, low_pressure):
+    if spatial_dimension == 2:
+      N = 64
+    else:
+      N = 128
+
+    if dtype == jnp.float32:
+      max_grad_thresh = 1e-5
+      atol = 1e-4
+      rtol = 1e-3
+    else:
+      max_grad_thresh = 1e-10
+      atol = 1e-8
+      rtol = 1e-5
+
+    for index in range(STOCHASTIC_SAMPLES):
+      cijkl, R, sigma, box = test_util.load_nc_data(spatial_dimension, low_pressure, dtype, index) 
+      R = space.transform(box, R)
+      box = box[0,0]
+
+      displacement, shift = space.periodic(box)
+      energy_fn = energy.soft_sphere_pair(displacement, sigma=sigma)
+      assert( jnp.max(jnp.abs(grad(energy_fn)(R))) < max_grad_thresh )
+
+      EMT_fn = jit(elasticity.AthermalElasticModulusTensor(energy_fn, check_convergence=True))
+      C, converged = EMT_fn(R,box)
+      assert( C.dtype == dtype )
+      assert( C.shape == (spatial_dimension,spatial_dimension,spatial_dimension,spatial_dimension) )
+      if converged:
```

Comment:
Should we `self.assertTrue` that the calculation converged? Otherwise, I could imagine a situation in which it doesn't converge but the test passes.

Reply:
Tell me if you disagree, but I think we *don't* want to assert that. When using float32's, it is possible that you simply don't have enough precision for convergence. This isn't a fault of the solver, its just a consequence of the inputs. The guarantee that we want to make is that *if* the solver converges, then the elastic constant calculation is correct, but if the solver doesn't converge we make no promises. In my testing on sphere packings, the solver always converges except for low pressure systems with float32 precision. 

If you are worried about this being dangerous, perhaps we should make the `check_convergence` flag default to `True`. That way someone has to actively turn it off, meaning hopefully they have read the docs and are aware of the risk. We could also be more strict by setting the answer to all `nan`s if the solver doesn't converge. I'm open to suggestions.